### PR TITLE
Add unit test for filesystem service.

### PR DESCRIPTION
**Pull request by dichenqiandbt**
## What is this PR?
This is a:
- [x] documentation update
- [ ] bug fix with no breaking changes
- [ ] new functionality
- [ ] a breaking change

All pull requests from community contributors should target the `main` branch (default).

## Description & motivation
One find is that for async end point, as we change our working directory, the dbt log file is stored under dbt project instead of dbt server(I think it's not expected).

## Checklist
- [ ] I have verified that these changes work locally on the following warehouses (Note: it's okay if you do not have access to all warehouses, this helps us understand what has been covered)
    - [ ] BigQuery
    - [ ] Postgres
    - [ ] Redshift
    - [ ] Snowflake
    - [ ] Databricks
    - [ ] Spark
- [ ] I have updated the README.md (if applicable)
- [ ] I have added tests & descriptions to my models
- [ ] I have added an entry to CHANGELOG.md

### CODE REVIEW

File: dbt_server/services/filesystem_service.py [hunk 1]
```diff
@@ -11,19 +11,51 @@
...
-def get_root_path(state_id=None, project_path=None):
+def get_root_path(state_id: str = None, project_path: str = None):
+    """Returns root path of dbt project. Only one of `state_id` and
+    `project_path` should be passed in. If both arguments are None, returns
+    None.
+
+    Args:
+        state_id: if set, we infer the dbt project is not stored locally,
+            instead it's pushed by client. Hence root path is a temp directory
+            that stores pushed dbt project.
```
Comment (racheldaniel) [inline dbt_server/services/filesystem_service.py:56]:
```suggestion
            instead it's pushed by client. Hence root path is a directory
            that stores the current state of the pushed dbt project.
```

File: dbt_server/services/filesystem_service.py [hunk 2]
```diff
@@ -93,13 +144,20 @@ def write_file(path, contents):
 
 
 @tracer.wrap
-def copy_file(source_path, dest_path):
-    ensure_dir_exists(dest_path)
+def copy_file(source_path: str, dest_path: str):
+    """Copies file from `source_path` to `dest_path`. The direcory of
```
Comment (racheldaniel) [inline dbt_server/services/filesystem_service.py:148]:
```suggestion
    """Copies file from `source_path` to `dest_path`. The directory of
```

File: dbt_server/services/filesystem_service.py [hunk 3]
```diff
@@ -93,13 +144,20 @@ def write_file(path, contents):
...
+    `dest_path` will be created recursively if not existed."""
```
Comment (racheldaniel) [inline dbt_server/services/filesystem_service.py:149]:
```suggestion
    """Copies file from `source_path` to `dest_path`. The directory of
    `dest_path` will be created recursively if it doesn't exist."""
```

File: dbt_server/services/filesystem_service.py [hunk 4]
```diff
@@ -108,7 +166,17 @@ def read_serialized_manifest(path):
...
+    state_id: str, previous_state_id: str, filedict: dict
+):
+    """Writes files in `filedict` to root path specified by `state_id`, then
+    copies previous partial parsed msgpack to current root path.
+
+    Args:
+        state_id: required to get root path.
+        previous_state_id: if it's none, we'll skip copy previous partial parsed
+            msgpack to current root path.
+        filedict: key is file name and value is file_info."""
```
Comment (racheldaniel) [inline dbt_server/services/filesystem_service.py:179]:
```suggestion
        filedict: key is file name and value is FileInfo."""
```

File: dbt_server/services/filesystem_service_test.py [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+from dbt_server.services.filesystem_service import get_path
+from dbt_server.services.filesystem_service import get_size
+from dbt_server.services.filesystem_service import write_file
+from dbt_server.services.filesystem_service import copy_file
+from dbt_server.services.filesystem_service import read_serialized_manifest
+from dbt_server.services.filesystem_service import write_unparsed_manifest_to_disk
+from dbt_server.services.filesystem_service import get_latest_state_id
+from dbt_server.services.filesystem_service import get_latest_project_path
+from dbt_server.services.filesystem_service import update_state_id
+from dbt_server.services.filesystem_service import update_project_path
```
Comment (racheldaniel) [inline dbt_server/services/filesystem_service_test.py:20]:
```suggestion
from dbt_server.services.filesystem_service import (
    get_working_dir,
    get_target_path,
    get_root_path,
    get_task_artifacts_path,
    get_log_path,
    get_partial_parse_path,
    get_db_path,
    get_latest_state_file_path,
    get_latest_project_path_file_path,
    get_path,
    get_size,
    write_file,
    copy_file,
    read_serialized_manifest,
    write_unparsed_manifest_to_disk,
    get_latest_state_id,
    get_latest_project_path,
    update_state_id,
    update_project_path,
)
```

File: dbt_server/services/filesystem_service_test.py [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+from os import makedirs
+from os import environ
+from os import path
```
Comment (racheldaniel) [inline dbt_server/services/filesystem_service_test.py:23]: For my own knowledge, is there a reason you choose to separate imports this way?

Comment (dichenqiandbt) [reply dbt_server/services/filesystem_service_test.py:23]: It's a pure code style decision and my reference is 
- Google python lanuage style https://google.github.io/styleguide/pyguide.html#313-imports-formatting
- Some stackflow discussion https://stackoverflow.com/questions/9916878/importing-modules-in-python-best-practice

I think the advantage to separate them out is it's the clearest and easiest understand way. Compare with
from xx import a, b, c(sometimes the import list is too long and hard to read) or from xx import (a,
b,
c
)(I feel like my eye needs to jump between lines).

That's definitely a code style thinking, I'm neutral on other solutions but lean to separate imports. Hope to align with you(then we can strictly following this later). Most of my python code style is the Google one, feel free to offline discuss with me.

File: dbt_server/services/filesystem_service_test.py (hunk 5 above)
Comment (dichenqiandbt) [reply dbt_server/services/filesystem_service_test.py:20]: See reply below.

File: dbt_server/services/filesystem_service.py [hunk 7]
```diff
@@ -11,19 +11,51 @@
 # This is defined in dbt-core-- dir path is configurable but not filename
 DBT_LOG_FILE_NAME = "dbt.log"
 
+# File system tree structure.
+# root_path(dbt project)
```
[the remark this answers is not in the document]
Comment (dichenqiandbt) [reply dbt_server/services/filesystem_service.py]: Thanks for pointing out! I just draw two different trees and hope it's clear.

File: dbt_server/services/filesystem_service_test.py (hunk 6 above)
Comment (racheldaniel) [reply dbt_server/services/filesystem_service_test.py:23]: Thanks for the info! I've never seen it this way but don't have a preference. My only hesitation is that we may want to keep our style consistent with [dbt-core](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/task/base.py#L13) for contribution reasons-- but I don't feel strongly about that either. Totally your call.

**Review by racheldaniel — APPROVED** [review]
Looks great, thank you!

**Pull request merged**